File: src/agent/agent-sdk/nodejs/src/download.ts
/*
 * BK-CI Agent SDK - 升级文件下载（worker.jar / docker init 脚本等）
 *
 * 对应 Go agent 的 src/pkg/util/httputil/devops.go 的 DownloadUpgradeFile 与
 * src/pkg/api/api.go 的 DownloadUpgradeFile / src/pkg/upgrade/download/*。
 *
 * 实现 eTag/md5 增量下载协议：
 *   - 本地已存在目标文件时，算其 md5 作为 &eTag=<md5> 传给后台；
 *   - 200 → 原子写入（写临时文件再 rename），并校验响应头 X-Checksum-Md5；
 *   - 304 → 文件未变更，返回旧 md5，notModified=true；
 *   - 404 → 抛错 file not found。
 *
 * 说明：与 Go 一致，下载走普通网关（getGateway()）而非 fileGateway，
 * 鉴权头用 config.getAuthHeaderMap()。零第三方依赖，基于 httpClient.requestStream。
 */

import * as fs from 'fs';
import * as path from 'path';
import * as crypto from 'crypto';
import { pipeline } from 'stream';
import { promisify } from 'util';
import { requestStream } from './httpClient';

const pipelineAsync = promisify(pipeline);

/** 下载接口路径前缀（对应 Go buildUrl 拼的 upgrade/files/download）。 */
export const DownloadApiPath =
  '/ms/environment/api/buildAgent/agent/thirdPartyAgent/upgrade/files/download';

/** worker.jar 在本地/服务端的固定文件名（对应 Go config.WorkAgentFile）。 */
export const WorkAgentFile = 'worker-agent.jar';
/** worker.jar 在服务端的下载相对路径（对应 Go "jar/worker-agent.jar"）。 */
export const WorkerJarServerFile = 'jar/worker-agent.jar';

/** docker init 脚本在本地的固定文件名（对应 Go config.DockerInitFile）。 */
export const DockerInitFile = 'agent_docker_init.sh';

export interface DownloadResult {
  /** 下载后（或未变更时的旧）文件 md5。 */
  md5: string;
  /** 是否为 304 未变更。 */
  notModified: boolean;
}

export interface DownloadFileOptions {
  /** 后台网关地址（可不带 http:// 前缀，会自动补全）。 */
  gateway: string;
  /** 鉴权头（config.getAuthHeaderMap()）。 */
  authHeaders: Record<string, string>;
  /** 服务端文件相对路径，作为 ?file= 查询参数，如 "jar/worker-agent.jar"。 */
  serverFile: string;
  /** 本地保存的绝对路径。 */
  savePath: string;
  /** 请求超时（毫秒），默认 300000（下载可能较大）。 */
  timeoutMs?: number;
}

function ensureGateway(gateway: string): string {
  if (gateway.startsWith('http://') || gateway.startsWith('https://')) {
    return gateway;
  }
  return 'http://' + gateway;
}

/** 计算文件 md5；文件不存在返回空串（对应 Go GetFileMd5 对不存在的处理）。 */
export async function fileMd5(filePath: string): Promise<string> {
  return new Promise((resolve, reject) => {
    let stream: fs.ReadStream;
    try {
      stream = fs.createReadStream(filePath);
    } catch (e) {
      resolve('');
      return;
    }
    const hash = crypto.createHash('md5');
    stream.on('error', (err: NodeJS.ErrnoException) => {
      if (err.code === 'ENOENT') {
        resolve('');
      } else {
        reject(err);
      }
    });
    stream.on('data', (chunk) => hash.update(chunk));
    stream.on('end', () => resolve(hash.digest('hex')));
  });
}

/**
 * 下载一个升级文件，实现 eTag/md5 增量协议。
 *
 * @returns 下载后文件 md5 与是否未变更。
 * @throws  404 时抛 "file not found"，md5 不匹配时抛 "file md5 not match"，其余非 2xx 抛错。
 */
export async function downloadFile(opts: DownloadFileOptions): Promise<DownloadResult> {
  const oldMd5 = await fileMd5(opts.savePath);

  let url =
    ensureGateway(opts.gateway) +
    DownloadApiPath +
    '?file=' +
    encodeURIComponent(opts.serverFile);
  if (oldMd5 !== '') {
    url += '&eTag=' + oldMd5;
  }

  const resp = await requestStream({
    method: 'GET',
    url,
    headers: opts.authHeaders,
    timeoutMs: opts.timeoutMs ?? 300_000,
  });

  // 非 2xx 的分支处理。注意：304 属于非 2xx，需先消费/丢弃响应流。
  if (!(resp.status >= 200 && resp.status < 300)) {
    resp.stream.resume(); // 丢弃响应体，释放 socket
    if (resp.status === 404) {
      throw new Error('file not found');
    }
    if (resp.status === 304) {
      return { md5: oldMd5, notModified: true };
    }
    throw new Error(`download file failed, status=${resp.status}, url=${opts.serverFile}`);
  }

  // 原子写入：先写临时文件再 rename（对应 Go AtomicWriteFile）。
  await fs.promises.mkdir(path.dirname(opts.savePath), { recursive: true });
  const tmpPath = opts.savePath + '.tmp.' + process.pid + '.' + Date.now();
  try {
    const out = fs.createWriteStream(tmpPath, { mode: 0o755 });
    await pipelineAsync(resp.stream, out);
    await fs.promises.rename(tmpPath, opts.savePath);
  } catch (e) {
    await fs.promises.rm(tmpPath, { force: true }).catch(() => undefined);
    throw e;
  }

  const newMd5 = await fileMd5(opts.savePath);

  const checksumMd5 = firstHeader(resp.headers['x-checksum-md5']);
  if (checksumMd5 && checksumMd5 !== newMd5) {
    throw new Error('file md5 not match');
  }

  return { md5: newMd5, notModified: false };
}

function firstHeader(v: string | string[] | undefined): string {
  if (v === undefined) return '';
  return Array.isArray(v) ? (v[0] ?? '') : v;
}

/**
 * 下载 worker.jar 到 <dir>/worker-agent.jar。
 * 对应 Go：serverFile="jar/worker-agent.jar"。
 */
export function downloadWorkerJar(
  gateway: string,
  authHeaders: Record<string, string>,
  dir: string,
  timeoutMs?: number
): Promise<DownloadResult> {
  return downloadFile({
    gateway,
    authHeaders,
    serverFile: WorkerJarServerFile,
    savePath: path.join(dir, WorkAgentFile),
    timeoutMs,
  });
}

/**
 * 下载 docker 初始化脚本到 <dir>/agent_docker_init.sh。
 * 服务端相对路径随平台不同（对应 Go download_unix/darwin/win.go）：
 *   - linux:  script/linux/agent_docker_init.sh
 *   - darwin: script/macos/agent_docker_init.sh
 *   - win32:  script/windows/agent_docker_init.sh
 */
export function downloadDockerInitFile(
  gateway: string,
  authHeaders: Record<string, string>,
  dir: string,
  platform: NodeJS.Platform = process.platform,
  timeoutMs?: number
): Promise<DownloadResult> {
  let serverFile: string;
  switch (platform) {
    case 'darwin':
      serverFile = 'script/macos/agent_docker_init.sh';
      break;
    case 'win32':
      serverFile = 'script/windows/agent_docker_init.sh';
      break;
    default:
      serverFile = 'script/linux/agent_docker_init.sh';
  }
  return downloadFile({
    gateway,
    authHeaders,
    serverFile,
    savePath: path.join(dir, DockerInitFile),
    timeoutMs,
  });
}

File: src/agent/agent-sdk/python/bkci_agent_sdk/download.py
"""Incremental and checksum-verified downloads for agent upgrade files."""

from __future__ import annotations

import asyncio
import hashlib
import os
from pathlib import Path
import shutil
import sys
import tempfile
from typing import Mapping, NamedTuple
from urllib.parse import urlencode

from .http_client import request_stream

DOWNLOAD_API_PATH = (
    "/ms/environment/api/buildAgent/agent/thirdPartyAgent/upgrade/files/download"
)
WORK_AGENT_FILE = "worker-agent.jar"
WORKER_JAR_SERVER_FILE = "jar/worker-agent.jar"
DOCKER_INIT_FILE = "agent_docker_init.sh"


class DownloadResult(NamedTuple):
    md5: str
    not_modified: bool


def _ensure_gateway(gateway: str) -> str:
    return gateway if gateway.startswith(("http://", "https://")) else "http://" + gateway


def _file_md5_sync(file_path: str | Path) -> str:
    digest = hashlib.md5()
    try:
        with Path(file_path).open("rb") as stream:
            for chunk in iter(lambda: stream.read(1024 * 1024), b""):
                digest.update(chunk)
    except FileNotFoundError:
        return ""
    return digest.hexdigest()


async def file_md5(file_path: str | Path) -> str:
    return await asyncio.to_thread(_file_md5_sync, file_path)


async def download_file(
    *,
    gateway: str,
    auth_headers: Mapping[str, str],
    server_file: str,
    save_path: str | Path,
    timeout_seconds: float = 300,
) -> DownloadResult:
    save = Path(save_path)
    old_md5 = await file_md5(save)
    query: dict[str, str] = {"file": server_file}
    if old_md5:
        query["eTag"] = old_md5
    url = f"{_ensure_gateway(gateway)}{DOWNLOAD_API_PATH}?{urlencode(query)}"
    response = await request_stream(
        method="GET", url=url, headers=auth_headers, timeout_seconds=timeout_seconds
    )

    if not 200 <= response.status < 300:
        await asyncio.to_thread(response.stream.close)
        if response.status == 404:
            raise FileNotFoundError("file not found")
        if response.status == 304:
            return DownloadResult(old_md5, True)
        raise RuntimeError(
            f"download file failed, status={response.status}, url={server_file}"
        )

    await asyncio.to_thread(save.parent.mkdir, parents=True, exist_ok=True)
    fd, temp_name = tempfile.mkstemp(prefix=save.name + ".tmp.", dir=save.parent)
    os.close(fd)
    temp_path = Path(temp_name)
    try:
        def copy_response() -> None:
            try:
                with temp_path.open("wb") as output:
                    shutil.copyfileobj(response.stream, output, 1024 * 1024)
            finally:
                response.stream.close()

        await asyncio.to_thread(copy_response)
        new_md5 = await file_md5(temp_path)
        checksum = (response.headers.get("X-Checksum-Md5") or "").strip()
        if checksum and checksum != new_md5:
            raise RuntimeError("file md5 not match")
        await asyncio.to_thread(os.chmod, temp_path, 0o755)
        await asyncio.to_thread(os.replace, temp_path, save)
        return DownloadResult(new_md5, False)
    finally:
        try:
            temp_path.unlink(missing_ok=True)
        except OSError:
            pass


async def download_worker_jar(
    gateway: str,
    auth_headers: Mapping[str, str],
    directory: str | Path,
    timeout_seconds: float = 300,
) -> DownloadResult:
    return await download_file(
        gateway=gateway,
        auth_headers=auth_headers,
        server_file=WORKER_JAR_SERVER_FILE,
        save_path=Path(directory) / WORK_AGENT_FILE,
        timeout_seconds=timeout_seconds,
    )


async def download_docker_init_file(
    gateway: str,
    auth_headers: Mapping[str, str],
    directory: str | Path,
    *,
    platform: str | None = None,
    timeout_seconds: float = 300,
) -> DownloadResult:
    target_platform = platform or sys.platform
    if target_platform == "darwin":
        server_file = "script/macos/agent_docker_init.sh"
    elif target_platform in {"win32", "cygwin"}:
        server_file = "script/windows/agent_docker_init.sh"
    else:
        server_file = "script/linux/agent_docker_init.sh"
    return await download_file(
        gateway=gateway,
        auth_headers=auth_headers,
        server_file=server_file,
        save_path=Path(directory) / DOCKER_INIT_FILE,
        timeout_seconds=timeout_seconds,
    )

